### PR TITLE
Improve SYCL reduction performance: workgroup_reduction

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #6035 taking care of the improvements to the `workgroup_reduction` functions used by all policies.
- It turned out that shuffle-based reductions are slower than going through local memory with the current implementation. This mirrors what we see for `Cuda` and `HIP` but it's not obvious why that is necessarily the case and whether we could improve the shuffle-based implementation some more. That's why I would like to keep the code around.
- I saw minor improvements when helping the compiler understand that `value_count == 1` for non-array reductions as well as using `int` for local indices.
- We can avoid copying intermediate results in one place if we just read from the appropriate place later. 
- We are running into some issues on `Volta70` on one test case, see https://github.com/kokkos/kokkos/pull/6270#discussion_r1262665349. Since we mostly care about Intel GPUs, and A100s are more widely available, this pull request switches testing to an A100.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 1]
```diff
@@ -30,61 +30,63 @@ namespace Kokkos {
...
       final_reducer.join(result, &local_mem[(local_id + stride) * value_count]);
     sycl::group_barrier(sg);
   }
   sycl::group_barrier(item.get_group());
 
-  // Copy the subgroup results into the first positions of the
-  // reduction array.
-  if (id_in_sg == 0)
-    final_reducer.copy(&local_mem[sg.get_group_id()[0] * value_count], result);
-  sycl::group_barrier(item.get_group());
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:67]: With all the integer type dance, it is hard to follow what happens to that block.
Why is it not needed anymore?

**Review by dalg24 — COMMENTED** [review]
What type of integer was used before?  `size_t`?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 2]
```diff
@@ -117,29 +119,34 @@ std::enable_if_t<use_shuffle_based_algorithm<ReducerType>> workgroup_reduction(
...
-  const auto max_subgroup_size = sg.get_max_local_range()[0];
-  const auto n_active_subgroups =
+  const int max_subgroup_size = sg.get_max_local_range()[0];
+  const unsigned int n_active_subgroups =
       (max_size + max_subgroup_size - 1) / max_subgroup_size;
-  if (id_in_sg == 0 && sg.get_group_id()[0] <= n_active_subgroups)
-    local_mem[sg.get_group_id()[0]] = local_value;
+  const int sg_group_id = sg.get_group_id()[0];
+  if (id_in_sg == 0 && sg_group_id <= static_cast<int>(n_active_subgroups))
+    local_mem[sg_group_id] = local_value;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:141]: I wanted a named variable and a constant loop bound. I could make it `unsigned int` or `n_active_subgroups` `int`.

File: .jenkins [hunk 3]
```diff
@@ -111,7 +111,7 @@ pipeline {
                                 -DCMAKE_CXX_FLAGS="-fsycl-device-code-split=per_kernel -Wno-deprecated-declarations -Werror -Wno-gnu-zero-variadic-macro-arguments -Wno-unknown-cuda-version -Wno-sycl-target" \
                                 -DKOKKOS_IMPL_SYCL_DEVICE_GLOBAL_SUPPORTED=0 \
                                 -DKokkos_ARCH_NATIVE=ON \
-                                -DKokkos_ARCH_VOLTA70=ON \
+                                -DKokkos_ARCH_AMPERE80=ON \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply .jenkins:114]: Is that something you could reproduce on another system?
If we merge this, are we implicitly saying that we no more "support" Volta?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:141]: This was my point.  If you make `n_active_subgroups` an `int` you don't need the cast on line 148 either

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:67]: The values are written in line 58 and read in lines 72, 81, and 89.
Previously, we copied them to the first positions in that array but that isn't really necessary. That means that we are not reading from consecutive memory locations anymore, but it seems to be better performance-wise this way.

File: .jenkins (hunk 3 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply .jenkins:114]: We never (officially) supported any Nvidia GPU anyway. 🙂 
It's failing in one case so we should still largely be able to use `Volta70`. 

I could try reproducing this on other machines but I don't have any with the compiler already there or access to `Docker`.

[the remark this answers is not in the document]
Comment (masterleinad) [reply .jenkins:114]: > you need to mention that change in the PR description.

Done.

**Comment by masterleinad** [comment]
SYCL CI is passing.
